### PR TITLE
i#6367: Insert FILTER_ENDPOINT when going to TRACE from COUNT also.

### PR DESCRIPTION
**Pull request by prasun3**
For BIMODAL traces (see https://github.com/DynamoRIO/dynamorio/issues/5199), the normal transition is from COUNT to L0_FILTER to TRACE. But some threads may transition from COUNT to TRACE directly e.g., if they were not scheduled during L0_FILTER mode. In this case, the FILTER_ENDPOINT marker is not inserted currently, which leads to raw2trace failures.

An example cmd line that reproduces the bug is shown below. Note that there is only one worker thread which transitions through L0_FILTER and TRACE mode while the main thread is still in COUNT mode.

```
bin64/drrun -t drcachesim -offline \
-trace_after_instrs 2200K -L0_filter_until_instrs 10K -max_trace_size 10K -- ./threadsig 1 1000
```

Issue: #6367

### CODE REVIEW

**Comment by prasun3** [comment]
The aarch64 failures seem related to this feature but I have no idea why this would fail only for aarch64.

The 'x86-64-ubuntu22' failure is `api.rseq`.

File: clients/drcachesim/tracer/output.cpp [hunk 1]
```diff
@@ -1100,8 +1100,10 @@ process_and_output_buffer(void *drcontext, bool skip_size_cap)
     // Switch to instruction-tracing mode by adding FILTER_ENDPOINT marker if another
     // thread triggered the switch.
     ptr_int_t mode = tracing_mode.load(std::memory_order_acquire);
-    if (get_local_mode(data) != mode) {
-        if (get_local_mode(data) == BBDUP_MODE_L0_FILTER) {
+    ptr_int_t local_mode = get_local_mode(data);
+    if (op_L0_filter_until_instrs.get_value() && local_mode != mode &&
+        mode == BBDUP_MODE_TRACE) {
+        if (local_mode == BBDUP_MODE_L0_FILTER || local_mode == BBDUP_MODE_COUNT) {
```
Comment (derekbruening) [inline clients/drcachesim/tracer/output.cpp:1106]: Is there any other possibility for `local_mode != mode`?  Turn into an assert?

Comment (derekbruening) [inline clients/drcachesim/tracer/output.cpp:1106]: Is it possible to have a test reliably hit this?  Or the timing is too hard to guarantee?  At least discuss in the PR description how you tested this.

**Review by derekbruening — APPROVED** [review]

File: clients/drcachesim/tracer/output.cpp [hunk 2]
```diff
@@ -1100,8 +1100,10 @@ process_and_output_buffer(void *drcontext, bool skip_size_cap)
     // Switch to instruction-tracing mode by adding FILTER_ENDPOINT marker if another
```
[the remark this answers is not in the document]
Comment (prasun3) [reply clients/drcachesim/tracer/output.cpp:1100]: Added.

[the remark this answers is not in the document]
Comment (prasun3) [reply clients/drcachesim/tracer/output.cpp:1100]: Sorry about the typo. Fixed.

File: clients/drcachesim/tracer/output.cpp (hunk 1 above)
Comment (prasun3) [reply clients/drcachesim/tracer/output.cpp]: I have added a test cmd line to the description. I tried adding the same test to the suite, but weirdly it produces no traces (and the test fails with `raw2trace failed: No thread files`). But when I copy paste the test cmd line (and replace the semi-colons with spaces) and run it on the terminal, it runs as expected.

Comment (prasun3) [reply clients/drcachesim/tracer/output.cpp]: I changed it to an assert.

**Comment by prasun3** [comment]
x86-64 failed with https://github.com/DynamoRIO/dynamorio/issues/6185 again

**Comment by prasun3** [comment]
x86-64-ubuntu22 failed with https://github.com/DynamoRIO/dynamorio/issues/6185 again

**Comment by prasun3** [comment]
x86-32 failed with #6303 

```
2023-10-19T03:16:51.7804322Z         Start 303: code_api|tool.drcachesim.opcode_mix
2023-10-19T03:16:51.7808645Z 303: Test command: /home/runner/work/dynamorio/dynamorio/build_debug-internal-32/bin32/drrun "-s" "90" "-quiet" "-debug" "-killpg" "-stderr_mask" "0xC" "-dumpcore_mask" "0" "-code_api" "-t" "drcachesim" "-ipc_name" "/home/runner/work/dynamorio/dynamorio/build_debug-internal-32/suite/tests/drtestpipe_opcode_mix" "-instr_encodings" "-simulator_type" "opcode_mix" "--" "/home/runner/work/dynamorio/dynamorio/build_debug-internal-32/suite/tests/bin/simple_app"
2023-10-19T03:16:51.7812271Z 303: Test timeout computed to be: 600
2023-10-19T03:16:51.7947961Z 303: Hello, world!
2023-10-19T03:16:51.7948379Z 303: Encoding size 2 != instr size 6
2023-10-19T03:16:51.7950291Z 303: drcachesim: /home/runner/work/dynamorio/dynamorio/clients/drcachesim/reader/reader.cpp:203: virtual bool dynamorio::drmemtrace::reader_t::process_input_entry(): Assertion `false' failed.
2023-10-19T03:16:51.7952833Z 303/414 Test #303: code_api|tool.drcachesim.opcode_mix ..............................***Failed  Required regular expression not found. Regex=[^Hello, world!
```

**Pull request merged**